### PR TITLE
Feature: Add reduce_average (weighted reduce_mean)

### PR DESCRIPTION
**Pull request by yaox12**
Add `tf.reduce_average` according to issue #7422

In this issue, some people doubt why this function is needed. I think broadcasting `weights` and corresponding `axis` is complex enough and worth a built-in function.

`np.average` is also contributed by me. Related [Issue](https://github.com/numpy/numpy/issues/10989) and [PR](https://github.com/numpy/numpy/pull/10994).  
I add support for any case where `len(axis) == rank(weights)`,  and it's the same in `tf.reduce_average` to keep compatibility with numpy.

A unit test is attached.

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 23 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 38 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 53 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 68 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 83 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 97 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yifeif** [comment]
Thanks for the PR @yaox12! Could you resolve the conflict?

@josh11b do you mind taking a look? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 20 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 49 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
